### ISSUE TITLE
We need a "signInFail" callback

### ISSUE THREAD

**Issue by gustavopch** [firebase/firebaseui-web#87]
There's a `signInSuccess` but not a `signInFail` callback. How can we handle the case when the user went to authentication but for some reason he comes back to the page unauthenticated?

**Comment by TMSCH** [comment]
If some error happens while the user tries to sign in, and comes back unauthenticated, FirebaseUI will handle it and display an error message, with some options for the user to try to sign in again.

As FirebaseUI handle errors, I am not sure to see a use case for such callback, but I may be missing some scenarios. How would you use it?

**Comment by gustavopch** [comment]
@TMSCH Here my use case:
- User opens my food ordering webapp, choose some food and fill some inputs.
- User clicks "Send order".
- If authenticated, order is sent, otherwise, a modal is opened with FirebaseUI social sign-in buttons saying "You need to sign in to send your order".
- The user clicks one of the buttons and is redirected to sign-in page.
- When the user comes back to my webapp, a loading spinner saying "Sending your order" is visible until `signInSuccess` callback runs and sends the order, but if something went wrong, I must do something instead of letting the loading spinner there forever.

**Comment by bojeil-google** [comment]
The main value for this is when there is no UI being displayed. All sign-in errors should be caught with a UI displayed (making signInFail not very compatible with the flow). If you discover an error that is not caught, please report it. There should not be a situation where the spinner is stuck forever, even network timeouts are caught). Instead of signInFail, you could add your own cancel/close button next to the sign in widget UI (calls ui.reset() to clear it underneath).

**Comment by gustavopch** [comment]
FirebaseUI currently is perfect for a common sign-in page.
But I'm trying to integrate the sign-in process seamlessly with the ordering process.
When he leaves my page to sign in, I leave a variable stored so that when he comes back I can show the spinner (and not the FirebaseUI widget).
The user doesn't think he's signing in. He's just ordering food and had to press "Continue with Google".
I need a `signInFail` even if it's only to hide the spinner and show the auth widget again so that the user can try with another sign-in method.
P.S.: the spinner I'm talking about is a custom spinner with a text "Sending order" in my page (not related to FirebaseUI).

**Comment by bojeil-google** [comment]
When the user comes back to your page, can't you add an firebase.auth().onAuthStateChanged listener to check the current auth state? if that triggers with null and your saved variable is set, you can determine the sign in attempt failed and remove your loading animation and display the desired UI.

**Comment by gustavopch** [comment]
Not really.
`firebase.auth().onAuthStateChanged` always runs with a null `user` at a first time before FirebaseUI loads even if it's going to run again a few secs latter with an actual `user`. Doesn't seem to be reliable when using FirebaseUI.

**Comment by bojeil-google** [comment]
Hey @gustaopch, can you provide a more detailed overview of your system design? It would help me provide more informed advice. I can't tell if your are opening a popup page with firebaseui rendered, redirecting to that page, when you are setting your variable, etc.

**Comment by gustavopch** [comment]
@bojeil-google Sure. I'm sorry, forget about the variable I talked about. The way I found to make it work is the following.

I'm doing an SPA where FirebaseUI widget is immediately mounted with the page but not shown.

- When the user first enters, he sees the order page where he can add food to the shopping cart and fill some order info.
- After that, he clicks a "Send order" button which will push the order to Firebase Database if he's authenticated. If he's not authenticated, a modal window will be shown on top of everything where FirebaseUI widget is mounted and configured with redirect flow.
- Then the user clicks one of the sign-in buttons "Continue with Google" or "Continue with Facebook" (by the way, I had to fork the project to change the buttons' texts; it'd be great to be able to customize without forking) and is redirected to the corresponding provider.
- When the user comes back (FirebaseUI widget is mounted but not shown), I need a way to check if he did in fact sign in (which means I must show the spinner and push the order to the database) or if something went wrong. As `firebase.auth().onAuthStateChanged` is not reliable because it always runs with a null `user` at a first moment (maybe because FirebaseUI takes a few secs to load?), I'm using `firebase.apps[1].auth().getRedirectResult` which is not very elegant but is reliable.

**Comment by TMSCH** [comment]
@gustavopch thanks for the further explanations. I understand what you want to achieve, however I think FirebaseUI is not meant to be used in such a way. FirebaseUI widget internally handles complex flows and also keep track of pending states, not displaying it even after rendering it may create weird situations where the user should actually perform some actions in the widget.

For instance, a sign-in could be successful from the provider (Google, Facebook) point of view, but require some more work from your app perspective: the user might try to sign in with a Facebook account of email address user@example.com, while they already have an account signed up through Google with the same email address. In that case, the widget will offer the option to link the accounts.

A potential solution would be to use the [`popup`](https://github.com/firebase/firebaseui-web#sign-in-flows) flow of FirebaseUI: you would render the widget on the page, and the user would perform the provider sign-in flow in a popup. As the widget would still be rendered in the background, you don't have to keep track of pending state, and the signInSuccess callback would eventually be called, or the first view of the widget would be displayed.

Would that work for you?

**Comment by gustavopch** [comment]
@TMSCH Thanks for the explanation. Am I safe with popup flow or is it possible that it doesn't work because the browser is configured to avoid popups or something like that?

**Comment by TMSCH** [comment]
Browsers supported by Firebase SDK will not block popups when a user interaction is done, which is the case here (as a click on the "Continue with xxx" triggers the popup opening), so you are safe with popup flow.

**Comment by gustavopch** [comment]
@TMSCH So I'll use the popup flow. Thanks.

**Comment by TMSCH** [comment]
You're welcome, glad we found a solution!

**Comment by derwaldgeist** [comment]
Stumbled upon this as well. For phone authentication, the widget shows a "Cancel" button, but there's no way to handle this button. If the user clicks on it, the widget just reloads. How can I react on the user pressing the cancel button?

**Comment by bojeil-google** [comment]
Cancel will typically lead you back to the nascar page to restart the flow. If you wish to add a cancel button to reset the widget you can always add your own in the surrounding container. You can call `ui.reset()` to clear the UI.

**Comment by derwaldgeist** [comment]
You mean `ui.reset()` will remove the cancel button?

**Comment by bojeil-google** [comment]
`ui.reset()` clears the UI. The  inside cancel button is not meant to clear the UI. In many cases, it is needed to go back to the previous screen allowing a user to correct a flow, in case they made a mistake, changed their mind regarding a sign in option, etc. You can display a different button or action trigger in the outside container where you would on click clear the UI and display the relevant screen.

**Comment by derwaldgeist** [comment]
Ok, thanks for the clarification. In my case this leads to a bit weird UI, since I am showing the widget inside a dialog window. So the user would expect that "Cancel" would close the window. And I already have a Cancel button at the top of the dialog, since it's running on a mobile device.

**Comment by bojeil-google** [comment]
The cancel button is more relevant when multiple providers are supported which FirebaseUI is optimized for. It would typically return the user back to the Nascar page. You must be using only phone authentication. You could just hide the internal custom button in that case, something like.
```
.firebaseui-id-page-phone-sign-in-start .firebaseui-id-secondary-link {
  display: none;
}
```
And set your dialog close button to close the window.

**Comment by derwaldgeist** [comment]
Ok, I see. Wasn't sure if setting the CSS would work, as some parts of the widget seem to be displayed in an iframe. Will try that. And yes, I am using only phone authentication, as it is a direct replacement for Twitter digits.

**Comment by dilizarov** [comment]
I ran into the same exact issue as @derwaldgeist in that I'm replacing Twitter Digits (It was so perfect :(... ) and it is odd that the Cancel button doesn't really do anything if you're just using Phone Auth. I read the above, so I understand why, but I think having a custom uiConfig option for the Phone Provider that allows you to customize what happens when that Cancel is pressed would go a long way. In my case, I'd like to simply do `ui.reset()` from there. For now, I've applied the CSS above by @bojeil-google.
